File: android/navigation/src/main/kotlin/io/rynbridge/navigation/NavigationModule.kt
package io.rynbridge.navigation

import io.rynbridge.core.*

class NavigationModule(provider: NavigationProvider) : BridgeModule {

    override val name = "navigation"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "push" to { payload ->
            val screen = payload["screen"]?.stringValue ?: ""
            val params = payload["params"]?.dictionaryValue
            val result = provider.push(screen, params)
            result.toPayload()
        },
        "pop" to { _ ->
            val result = provider.pop()
            result.toPayload()
        },
        "popToRoot" to { _ ->
            val result = provider.popToRoot()
            result.toPayload()
        },
        "present" to { payload ->
            val screen = payload["screen"]?.stringValue ?: ""
            val style = payload["style"]?.stringValue
            val params = payload["params"]?.dictionaryValue
            val result = provider.present(screen, style, params)
            result.toPayload()
        },
        "dismiss" to { _ ->
            val result = provider.dismiss()
            result.toPayload()
        },
        "openURL" to { payload ->
            val url = payload["url"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: url")
            val result = provider.openURL(url)
            result.toPayload()
        },
        "canOpenURL" to { payload ->
            val url = payload["url"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: url")
            val result = provider.canOpenURL(url)
            result.toPayload()
        },
        "getInitialURL" to { _ ->
            val result = provider.getInitialURL()
            result.toPayload()
        },
        "getAppState" to { _ ->
            val result = provider.getAppState()
            result.toPayload()
        }
    )
}

File: ios/Sources/RynBridgeNavigation/NavigationModule.swift
import Foundation
import RynBridge

public struct NavigationModule: BridgeModule, Sendable {
    public let name = "navigation"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: NavigationProvider) {
        actions = [
            "push": { payload in
                let screen = payload["screen"]?.stringValue ?? ""
                let params = payload["params"]?.dictionaryValue
                let result = try await provider.push(screen: screen, params: params)
                return result.toPayload()
            },
            "pop": { _ in
                let result = try await provider.pop()
                return result.toPayload()
            },
            "popToRoot": { _ in
                let result = try await provider.popToRoot()
                return result.toPayload()
            },
            "present": { payload in
                let screen = payload["screen"]?.stringValue ?? ""
                let style = payload["style"]?.stringValue
                let params = payload["params"]?.dictionaryValue
                let result = try await provider.present(screen: screen, style: style, params: params)
                return result.toPayload()
            },
            "dismiss": { _ in
                let result = try await provider.dismiss()
                return result.toPayload()
            },
            "openURL": { payload in
                guard let url = payload["url"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: url")
                }
                let result = try await provider.openURL(url: url)
                return result.toPayload()
            },
            "canOpenURL": { payload in
                guard let url = payload["url"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: url")
                }
                let result = try await provider.canOpenURL(url: url)
                return result.toPayload()
            },
            "getInitialURL": { _ in
                let result = try await provider.getInitialURL()
                return result.toPayload()
            },
            "getAppState": { _ in
                let result = try await provider.getAppState()
                return result.toPayload()
            },
        ]
    }
}

#if canImport(UIKit)
extension NavigationModule {
    public init() {
        self.init(provider: DefaultNavigationProvider())
    }
}
#endif
